NetBox 3.X

snaselj opened the pull request:
# Closes: NaN

## What's Changed

- Dropped NetBox 2.x support.
- Dropped ChangeLog import.
- Implemented NetBox `3.0` - `3.6` support.
- Dropped Nautobot 1.x support.
- Implemented Nautobot `2.0` - `2.1` support.
- Added [importer development documentation](https://github.com/nautobot/nautobot-plugin-netbox-importer/blob/u/snaselj-netbox3/docs/dev/generator.md).

## Considerations

- [x] Implement writing `created` and `last_updated` fields.
- [x] Test to re-run the import.
- [x] Parametrize importer to fail if any validation fails.
- [x] Resolve skipped fields and TBDs in `nautobot_netbox_importer/diffsync/netbox.py`.
- [x] Resolve validation error: `django.core.exceptions.ValidationError: ['Rack R308 (Row 3) and power panel Panel 4 (MDF) are in different locations']`.
- [x] Resolve validation error: `django.core.exceptions.ValidationError: {'parent': ['A Location of type Location may only have a Location of the same type as its parent.']}`.
- [x] Change `InternalTypeStr` to `enum`.
- [x] Stream processing of input JSON file, to avoid loading the entire file into memory.

## Not a scope of this PR

- Implement mysql tests?
- Implement SSoT Job.
    - Allow DiffSync deleting in Nautobot?
    - Add-only `ManyToMany` field values?
    - Add-only custom field values?

glennmatthews commented:
Was it necessary to move so much code around as part of this feature? I don't really know where to begin reviewing this when diff-wise it basically looks like an entirely new implementation of the app.

glennmatthews commented:
Also, does this just *add* 3.0 support or does it also *drop support* for all older NetBox versions? I can't really tell with the scope of the diffs here.

snaselj commented:
> Was it necessary to move so much code around as part of this feature? I don't really know where to begin reviewing this when diff-wise it basically looks like an entirely new implementation of the app.

Yes, the implementation is different, supports multiple NetBox 3.0, 3.1 and 3.2 now, going to add support up to 3.6.

snaselj commented:
> Also, does this just _add_ 3.0 support or does it also _drop support_ for all older NetBox versions? I can't really tell with the scope of the diffs here.

Dropped support for Nautobot 1.x and NetBox 2.x, updated the description accordingly.

glennmatthews commented:
Just a quick note that changelog imports was a user-requested feature in the past. How difficult would it be to re-add?

glennmatthews commented on nautobot_netbox_importer/diffsync/base.py:
```diff
@@ -0,0 +1,132 @@
+            raise ValueError(f"Invalid content type {content_type}")
+        content_type = ".".join(content_type)
+
+    instance = ContentType.objects.get_for_model(get_model_from_name(content_type))
```
Might be more efficient to directly do something like `ContentType.objects.get(app_label=content_type[0], model=content_type[1])`?

snaselj replied:
Implemented better Nautobot content types handling including caching in `NautobotModelWrapper`.

glennmatthews commented on nautobot_netbox_importer/diffsync/base.py:
```diff
@@ -0,0 +1,132 @@
+import datetime
+import decimal
+import logging
+from typing import Any
+from typing import Iterable
+from typing import List
+from typing import Mapping
+from typing import MutableMapping
+from typing import Tuple
+from typing import Type
+from typing import Union
+from uuid import NAMESPACE_DNS
+from uuid import UUID
+from uuid import uuid5
```
nit: wouldn't the more usual black/isort/ruff convention be to group these, a la:

```suggestion
from typing import (
    Any,
    Iterable,
    List,
    Mapping,
    MutableMapping,
    Tuple,
    Type,
    Union,
)
from uuid import NAMESPACE_DNS, UUID, uuid5
```

Same question/comment throughout.

snaselj replied:
This way passes through all CI checks and produces better git diffs.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+NautobotFields = MutableMapping[FieldName, InternalFieldTypeStr]
+
+# Helper to get the order of models to import. Not all models needs to be here.
+IMPORT_ORDER: Iterable[ContentTypeStr] = (
```
What's the plan for maintaining/extending this as Nautobot adds new data models? Will it need to drop support for non-latest versions of Nautobot?

snaselj replied:
Cleaned up and added better description to `IMPORT_ORDER`. It's possible to have currently unsupported content types here, so we can add future content types with keeping old versions support.

The issue will arise once older Nautobot version will need the different order than a newer one.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+
+NautobotFields = MutableMapping[FieldName, InternalFieldTypeStr]
+
+# Helper to get the order of models to import. Not all models needs to be here.
```
Can you elaborate more on "Not all models needs to be here" -- what models do need to be here? what models don't? Why?

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+        """Get the Nautobot content type ID for a given content type."""
+        return ContentType.objects.get_for_model(get_model_from_name(self.content_type))
+
+    def set_instance_defaults(self, **defaults: Any) -> None:
+        """Set default values for a Nautobot instance constructor."""
+        self.constructor_kwargs = defaults
```
Would it make sense for this to be set at `__init__` time, or does it need to be a separate setter like this?

snaselj replied:
The structure of wrappers is created on the fly. Some parts can be set in the constructors, however it's possible, that the wrapper will already exists, before configuring it.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+
+    def get_content_type(self) -> ContentType:
+        """Get the Nautobot content type ID for a given content type."""
+        return ContentType.objects.get_for_model(get_model_from_name(self.content_type))
```
Same comment here about getting a ContentType directly.

Also, is this something that can/should be cached, perhaps at `__init__` time?

snaselj replied:
Implemented better Nautobot content types handling including caching in `NautobotModelWrapper`.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+                try:
+                    instance.clean()
+                except ValidationError as error:
+                    if errors:
+                        errors.add(error)
+                    else:
+                        errors = set([error])
+                        self.validation_errors[model_type._meta.label.lower()] = errors  # type: ignore
```
```suggestion
                    self.validation_errors.setdefault(model_type._meta.label.lower(), set()).add(error)
```

Would probably also be more efficient to cache that `model_type._meta.label.lower()` key instead of re-constructing it for each error.

snaselj replied:
Forgot about `setdefault()` already, good point.

Actually, that's what the current implementation does, converts the label only once.

snaselj replied:
Nvm, refactored using `setdefault()`

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+    def save_nautobot_instance(self, instance: NautobotBaseModel, values: RecordData, fields: NautobotFields) -> None:
+        """Save a Nautobot instance."""
+
+        def set_custom_field(value: Any):
+            custom_field_data = getattr(instance, "custom_field_data", None)
+            if custom_field_data is None:
+                raise TypeError("Missing custom_field_data")
+            custom_field_data.clear()
+            if value:
+                custom_field_data.update(value)
```
So is this is to set all of custom_field_data in one pass, not to set an individual custom field value? If so, might I suggest calling it `set_custom_field_data()` instead?

snaselj replied:
Renamed as suggested

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+            if value:
+                custom_field_data.update(value)
+
+        def set_generic_foreign_key(field: GenericForeignKey, value: Any):
```
Looks like `value` is expected to be a two-tuple, not Any?

snaselj replied:
Fixed typing.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+    def save_nautobot_instance(self, instance: NautobotBaseModel, values: RecordData, fields: NautobotFields) -> None:
+        """Save a Nautobot instance."""
+
+        def set_custom_field(value: Any):
```
Looks like `value` is expected to be a Mapping, not Any.

snaselj replied:
Fixed typing.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+                    setattr(instance, field_name, "")
+                else:
+                    setattr(instance, field_name, None)
+            else:
+                setattr(instance, field_name, value)
```
Seems weird for `set_empty` to have an else case for handling non-empty values?

snaselj replied:
Fixed.

glennmatthews commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,362 @@
+                uid = instance.id
+                if not isinstance(uid, (UUID, str, int)):
+                    raise TypeError(f"Invalid uid {uid}") from exc
+                if instance.__class__ in self.clean_failures:
+                    self.clean_failures[instance.__class__].add(uid)
+                else:
+                    self.clean_failures[instance.__class__] = set([uid])
```
```suggestion
                self.clean_failures.setdefault(instance.__class__, set()).add(uid)
```

snaselj replied:
Fixed.

glennmatthews commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,505 @@
+def _name_from_wrapper(wrapper) -> str:
+    """Return model name with upper cased first letter.
+
+    e.g. `dcim.site` -> `Site`
```
Any concern about cases where this isn't "correct", e.g. `VLAN`, `IPAddress`?

snaselj replied:
Removed not to be confusing. It's used for location types only.

snaselj commented:
> Just a quick note that changelog imports was a user-requested feature in the past. How difficult would it be to re-add?

I don't have any testing data for it, so it's really difficult to estimate. Depends on the level of correctness, if it's just about importing JSON change object without change, then it should be quite easy to implement. When the importer should convert fields in JSON based on the content type, some issues can arise.

glennmatthews commented on nautobot_netbox_importer/diffsync/source.py:
```diff
@@ -0,0 +1,940 @@
+                result = self.extend_wrapper.get_pk_from_uid(uid)
+                self.uid_to_pk_cache[uid] = result
+                return result
+            return source_pk_to_uuid(self.content_type or self.content_type, uid)
```
Why don't we set `self.uid_to_pk_cache` in this case?

chadell commented:
Overall feedback from notes:

- Dropped NetBox 2.x support. -> Then, we should release a netbox-import version that supports as much as we did previous to this PR, and add a compatibility matrix
- Dropped ChangeLog import. -> We should add this back, as @glennmatthews said, it was a customer ask and I see value for adopters.
- Implemented NetBox 3.0 - 3.4 support.
- Dropped Nautobot 1.x support. -> The same as point 1, we should have a release of NetBox 2.x -> Nautobot 1.x
- Implemented Nautobot 2.0 support.

chadell commented on docs/dev/importer.md:
```diff
@@ -0,0 +1,154 @@
+
+### Defining the Source Structure Deviations
+
+Before importing, it is essential to define any deviations between the source structure and the target Nautobot structure. This is configured within `nautobot_plugin_netbox_importer/diffsync/netbox.py`.
```
`the source` == `the source NetBox`, isn't it? 
as we are using, `the target Nautobot`, I read the second easier

chadell commented:
Regarding "Implement SSoT Job", what's the approach?

chadell commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,497 @@
+            "created",
+            "last_updated",
+        ),
+        ignore_content_types={
```
if we define the configured mappings, couldn't we infer this? (due omission)

snaselj replied:
That's a good idea, will implement it.

snaselj replied:
This will be implemented in a separate pull request.

snaselj replied:
Skipping missing Nautobot models is implemented now.

chadell commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,497 @@
+            "contenttypes.contenttype": None,  # Nautobot has own content types
+            "sessions.session": None,  # Nautobot has own sessions
+            "dcim.cablepath": None,  # Recreated in Nautobot
+            "admin.logentry": None,  # NetBox 3.0 TBD verify
```
what's the strategy to handle different Netbox versions?

snaselj replied:
For now, we support 3.0 - 3.4, adding 3.5 and 3.6. The importer is created dynamically based on the current import data and Nautobot version. Till now, it's sufficient enough. If it will not work correctly, we can add versions constraints to `configure_model`, so that configuration will be applied only for particular Nautobot / NetBox versions.

chadell commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,497 @@
+            "users.userconfig": None,  # NetBox 3.0 TBD verify
+            "auth.permission": None,  # NetBox 3.0 TBD verify
+            "dcim.sitegroup": None,  # NetBox 3.1 TBD verify
+            "ipam.asn": None,  # NetBox 3.1 TBD verify
```
is it in the scope of this PR to leverage Nautobot apps (e.g., BGP) that could accommodate this model? and if so, recommend to install if not?

snaselj replied:
Nope, it's the future work.

chadell replied:
ack. Just to keep it in mind in the design

chadell commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,497 @@
+            "ipam.iprange": None,  # NetBox 3.1 TBD verify
+            "tenancy.contactgroup": None,  # NetBox 3.1 TBD verify
+            "tenancy.contactrole": None,  # NetBox 3.1 TBD verify
+            "tenancy.contact": None,  # NetBox 3.1 TBD verify
```
as this is coming soon to Nautobot, how would we accommodate different options for Nautobot?

snaselj replied:
Will implement inferred skipping of models, not present in Nautobot, so this configuration will go away.

snaselj commented:
> Overall feedback from notes:
> 
> * Dropped NetBox 2.x support. -> Then, we should release a netbox-import version that supports as much as we did previous to this PR, and add a compatibility matrix
> * Dropped ChangeLog import. -> We should add this back, as @glennmatthews said, it was a customer ask and I see value for adopters.
> * Implemented NetBox 3.0 - 3.4 support.
> * Dropped Nautobot 1.x support. -> The same as point 1, we should have a release of NetBox 2.x -> Nautobot 1.x
> * Implemented Nautobot 2.0 support.

Suppose the last release of this package is NetBox 2.x => Nautobot 1.x, is that right @whitej6 ? Or is should I create 1.x branches and another release?

snaselj commented:
> Regarding "Implement SSoT Job", what's the approach?

It's on pause now, depends on architecture decision.

chadell commented:
> > Overall feedback from notes:
> > 
> > * Dropped NetBox 2.x support. -> Then, we should release a netbox-import version that supports as much as we did previous to this PR, and add a compatibility matrix
> > * Dropped ChangeLog import. -> We should add this back, as @glennmatthews said, it was a customer ask and I see value for adopters.
> > * Implemented NetBox 3.0 - 3.4 support.
> > * Dropped Nautobot 1.x support. -> The same as point 1, we should have a release of NetBox 2.x -> Nautobot 1.x
> > * Implemented Nautobot 2.0 support.
> 
> Suppose the last release of this package is NetBox 2.x => Nautobot 1.x, is that right @whitej6 ? Or is should I create 1.x branches and another release?

IIRC, in `develop` there are some enhancements towards NetBox 2.11 support that were not released because not completed. I think it's valuable to get them out, even not completed.

whitej6 commented:
> > Regarding "Implement SSoT Job", what's the approach?
> 
> It's on pause now, depends on architecture decision.

In a later release an SSoT job will be added to help facilitate nautobot cloud users without requiring additional dev work in nautobot cloud to account for a uploading a file and running an arbitrary management command. It _may_ need a custom wrapper on the FileVar item as the field is stored in the database & we should avoid storing files that are Gigs in size in the DB. Great catch by @glennmatthews

whitej6 commented on pyproject.toml:
```diff
@@ -28,8 +28,8 @@ packages = [
 # Used for local development
-nautobot = "^1.6.0"
-diffsync = "^1.5.1"
+nautobot = "^2.0.0"
```
@glennmatthews would we want to have the minimum supported version be 2.0.6 due to the recent vulnerabilities & we haven't done a v2 supported release yet?

glennmatthews replied:
Good question. I'd be in favor of that as a guideline across the apps to be honest :-)

snaselj replied:
Updated minimum Nautobot version to 2.0.6

snaselj commented on nautobot_netbox_importer/tests/test_sync.py:
```diff
@@ -0,0 +1,155 @@
+    def test_sync_3_4(self):
+        self._sync("3.4")
+
+    def _sync(self, version: str):
```
Some models can be counted this way. However there is quite a lot of exceptions:

- All skipped models.
- Roles and Statuses.
- `dcim.cabletermination`

snaselj commented:
This PR is ready to go from my perspective. Will add `Considerations`, from the pull request description, as follow up tickets.

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+        "dcim.interface",
+        fields={
+            "status": "status",
+            "parent": None,  # TBD verify
```
```suggestion
            "parent": "parent_interface", 
```

snaselj replied:
Fixed

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+    adapter.configure_model(
+        "dcim.devicetype",
+        fields={
+            "front_image": None,  # TBD verify
```
These exist on both sides, I am assuming since it is an image it's not part of the JSON blog. If that is the case it's okay to document and move on.

snaselj replied:
Do not have data to test these for now, it's always empty in the available input. Will create follow-up ticket to cover this.

whitej6 replied:
May be easier to export a nautobot dataset that contains the front and back images to verify. I am 99% certain it's using django storages so they would not be in the database dump.

snaselj replied:
Confirming, it’s represented in the NetBox export JSON data as follows: 
```
    "front_image": "devicetype-images/screenshot.png",
```

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+        "virtualization.vminterface",
+        fields={
+            "status": "status",
+            "parent": None,  # TBD verify
```
```suggestion
            "parent": "parent_interface",
```

snaselj replied:
Fixed

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+        identifiers=["username"],
+        fields={
+            "last_login": None,
+            "is_superuser": None,  # NetBox 3.1 TBD verify
```
is_staff and is_superuser should match what we did in the previous version

snaselj replied:
Created follow-up ticket for this.

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+        fields={
+            "last_login": None,
+            "is_superuser": None,  # NetBox 3.1 TBD verify
+            "password": None,  # NetBox 3.1 TBD verify
```
Should not be attempted to migrated

snaselj replied:
Fixed the comment

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+            "last_login": None,
+            "is_superuser": None,  # NetBox 3.1 TBD verify
+            "password": None,  # NetBox 3.1 TBD verify
+            "user_permissions": None,  # NetBox 3.1 TBD verify
```
Permissions are a M2M from ObjectPermission to User not the reverse

snaselj replied:
Created follow-up ticket for this.

whitej6 commented on nautobot_netbox_importer/diffsync/netbox.py:
```diff
@@ -0,0 +1,491 @@
+        "auth.group",
+        identifiers=["name"],
+        fields={
+            "permissions": None,  # NetBox 3.1 TBD verify
```
Permissions are a M2M from ObjectPermission to Group not the reverse

snaselj replied:
Created follow-up ticket for this.

whitej6 commented on nautobot_netbox_importer/diffsync/nautobot.py:
```diff
@@ -0,0 +1,432 @@
+    "dcim.rearport",
+    "dcim.rearporttemplate",
+    "dcim.cable",
+    "ipam.prefix",
```
Do you need to add the remainder of IPAM?

snaselj replied:
Not necessary to add other models for now, this list is just to precede those models, that needs to be imported between others in a specific order.

whitej6 commented:
To be a stickler on this since diffsync is a little more advanced topic and I find sticking to the folder patters helpful with understanding where to go for certain files etc especially since models & adapters are somewhat circular for imports/use

`diffsync/adapters/<adapter name>.py`
`diffsync/models/base.py`
`diffsync/models/<adapter name>.py`

Keeping DiffSync & SSoT structures the same across multiple repos help with maintainability.

snaselj commented:
> To be a stickler on this since diffsync is a little more advanced topic and I find sticking to the folder patters helpful with understanding where to go for certain files etc especially since models & adapters are somewhat circular for imports/use
> 
> `diffsync/adapters/<adapter name>.py` `diffsync/models/base.py` `diffsync/models/<adapter name>.py`
> 
> Keeping DiffSync & SSoT structures the same across multiple repos help with maintainability.

I understand reasons, however this implementation needs tight coupling between source classes (adapter / model wrapper / field) as well between nautobot classes (adapter / model wrapper / field). To split the code the way you propose, mean significant refactoring, and actually a totally different implementation.

snaselj commented:
> > To be a stickler on this since diffsync is a little more advanced topic and I find sticking to the folder patters helpful with understanding where to go for certain files etc especially since models & adapters are somewhat circular for imports/use
> > `diffsync/adapters/<adapter name>.py` `diffsync/models/base.py` `diffsync/models/<adapter name>.py`
> > Keeping DiffSync & SSoT structures the same across multiple repos help with maintainability.
> 
> I understand reasons, however this implementation needs tight coupling between source classes (adapter / model wrapper / field) as well between nautobot classes (adapter / model wrapper / field). To split the code the way you propose, mean significant refactoring, and actually a totally different implementation.

Moved the code, to be better aligned with the current practice.

chadell commented on docs/user/app_use_cases.md:
```diff
@@ -4,83 +4,118 @@ This document describes common use-cases and scenarios for this App.
+#### Importing Change Log into Nautobot
 
-![Screenshot of the completed process](../images/screenshot2.png)
+From within the Nautobot application environment, run `nautobot-server import_netbox <json_file_only_objectchanges>`, for example, `nautobot-server import_netbox /tmp/netbox_objectchange.json`. This will skip all object change records that are not possible to map to their corresponding objects.
```
provide a n example as done above

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+update: 1
+delete: 0
+no-change: 4
+skip: 119
```
why thinks are skipped? and updated?

snaselj replied:
Described.

chadell commented on docs/user/summary.md:
````diff
@@ -0,0 +1,1115 @@
+
+```
+- Validation issues: -----------------------------------------------------------
+. dcim.powerfeed: 48 ...........................................................
````
you could trim a bit the verbose output to make it more readable

snaselj replied:
Trimmed.

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+dcim.modulebaytemplate: Nautobot Model Not Found
+dcim.moduletype: Nautobot Model Not Found
+dcim.module: Nautobot Model Not Found
+ipam.asn: Nautobot Model Not Found
```
should we point towards the BGP app? is it possible to adapt to it?

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+- Content Types Back Mapping: --------------------------------------------------
+  Back mapping deviations from Nautobot content type to the source content type
+extras.role => Ambiguous
+users.user => auth.user
```
why is this in this list if not "Ambiguous"?

snaselj replied:
This list contain all deviations, mapping to the different content type is considered a deviation.

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+
+## Content Types Back Mapping
+
+This section shows back mapping deviations from Nautobot content types to NetBox content types. Only those content types that differ between NetBox and Nautobot are shown.
```
what is the action task for the user with this info?

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+_name (DATA) => _name (PrivateProperty)
+label (DATA) => label (CharField)
+description (DATA) => description (CharField)
+module (DATA) => module (NotFound)
```
what does it mean? is this information lost?

snaselj replied:
Yes, `NotFound` fields are skipped as described above.

https://github.com/nautobot/nautobot-app-netbox-importer/pull/110/files#diff-1a4b92f2e8bf12ae4a44e4d63c8a42caa969e6f492a4b61e1ae19348b2841921R245

chadell commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+    - `(ReadOnlyProperty')` marks fields that are read-only properties in Nautobot and can't be imported.
+    - `(PrivateProperty)` marks fields that are prefixed with an underscore `_` and are considered private properties. Those fields are not imported.
+    - `(NotFound)` indicates the field is not found in Nautobot and can't be imported.
+    - `(DoNotImportLastUpdated)` are fields that are not imported as they are automatically updated by Nautobot.
```
the word seems only for "LastUpdated" not for any field updated by Nautobot

snaselj replied:
Updated the documentation to address this.

chadell commented on docs/dev/generator.md:
```diff
@@ -0,0 +1,201 @@
+
+The NetBox specific component is further segmented into sections, located in the `nautobot_netbox_importer/diffsync` directory:
+
+- `adapters/nautobot.py`: Inherits `NautobotAdapter` from the generic `NautobotAdapter` and implements pieces necessary for SSoT job.
```
why do we need a generic NautobotAdapter to create yet another NautobotAdapter? what is this abstraction providing?

snaselj replied:
The first one is DiffSync only related implementation. The other one is adding SSoT related things, but it's not probably necessary.

The reason to create the second one was to make the file structure the same way as is in other SSoT apps / integrations.

chadell commented on docs/dev/generator.md:
```diff
@@ -0,0 +1,201 @@
+
+The initial step involves creating a `SourceAdapter()`. It accepts an argument, `get_source_data`, which is `Callable` that returns `Iterable` of the source data items. Each source data item is encapsulated in `SourceRecord(content_type: ContentTypeStr, data: Dict)` instances. `SourceAdapter` constructor also passes any additional arguments to its ancestor, the `DiffSync` class.
+
+The data undergoes two cycles: the first to establish the structure and the second to import the actual data.
```
could you provide a structure with these two cycles and the sub steps involved?

snaselj replied:
Steps are described in this document, updated this doc to mention that.

chadell commented:
General Questions:
- What has to be done (as a developer) to support a Nautobot minor version?
- What has to be done (as a developer) to support a Nautobot major version?
- - What has to be done (as a developer) to support a Nautobot App?
- What has to be done (as a developer) to support a NetBox new version?
- What has to be done (as a user) to provide some customization in the import method?

bryanculver commented:
> General Questions:
> 
> * What has to be done (as a developer) to support a Nautobot minor version?
> * What has to be done (as a developer) to support a Nautobot major version?
> * What has to be done (as a developer) to support a Nautobot App?
> * What has to be done (as a developer) to support a NetBox new version?
> * What has to be done (as a user) to provide some customization in the import method?

I'd suggest more targeted questions & answers, because if nothing changed between versions it can be a noop (AFAIK the way this refactor works is that it's data-type check instead of version check alleviates many of those concerns). So:

* What has to be done (as a developer) if NetBox changes a field's data type?
* (same as above but Nautobot side)
* What has to be done (as a developer) if NetBox adds a new model?
* What has to be done (as a developer) if NetBox changes a relation from 1-to-many to many-to-many?

Link to places where we are already handling these changes instead of speaking in hypotheticals.

I'd also like to see two other pieces of information:
- A logical (not technical) diagram for the data from in from NetBox and out to Nautobot
    - [NetBox] -> `SourceAdapter` -> (what does data look like here and where it is "held") -> `TargetAdapter` -> [Nautobot]
    - And what types of decision trees does a data type transformation look like?
- A few examples of roadblocks we ran into with the existing architecture when adding support for new versions (the below are hypotheticals so they are likely incorrect, but for demonstrations for justification)
    - "When adding support for Netbox 3.X+1 we were copying and pasting much of the same code for 3.X"
    - "When adding support for Nautobot 2.0, copy and paste code for Netbox 3.2-3.6 had to have repetitive updates and we found bugs where it was fixed in one place and not others"
    - "Removing support for Nautobot 1.X was proving to be very difficult and was effectively a rewrite anyway"

snaselj commented:
> > General Questions:
> > 
> > * What has to be done (as a developer) to support a Nautobot minor version?
> > * What has to be done (as a developer) to support a Nautobot major version?
> > * What has to be done (as a developer) to support a Nautobot App?
> > * What has to be done (as a developer) to support a NetBox new version?
> > * What has to be done (as a user) to provide some customization in the import method?
> 
> I'd suggest more targeted questions & answers, because if nothing changed between versions it can be a noop (AFAIK the way this refactor works is that it's data-type check instead of version check alleviates many of those concerns). So:
> 
> * What has to be done (as a developer) if NetBox changes a field's data type?
> * (same as above but Nautobot side)
> * What has to be done (as a developer) if NetBox adds a new model?
> * What has to be done (as a developer) if NetBox changes a relation from 1-to-many to many-to-many?
> 
> Link to places where we are already handling these changes instead of speaking in hypotheticals.
> 
> I'd also like to see two other pieces of information:
> 
> * A logical (not technical) diagram for the data from in from NetBox and out to Nautobot

Still WIP

>   
>   * [NetBox] -> `SourceAdapter` -> (what does data look like here and where it is "held") -> `TargetAdapter` -> [Nautobot]
>   * And what types of decision trees does a data type transformation look like?
> * A few examples of roadblocks we ran into with the existing architecture when adding support for new versions (the below are hypotheticals so they are likely incorrect, but for demonstrations for justification)
>   
>   * "When adding support for Netbox 3.X+1 we were copying and pasting much of the same code for 3.X"
>   * "When adding support for Nautobot 2.0, copy and paste code for Netbox 3.2-3.6 had to have repetitive updates and we found bugs where it was fixed in one place and not others"
>   * "Removing support for Nautobot 1.X was proving to be very difficult and was effectively a rewrite anyway"

Questions are addressed in `doc/dev/faq.md` and `doc/user/faq.md`.

snaselj commented:
> General Questions:
> 
> * What has to be done (as a developer) to support a Nautobot minor version?
> * What has to be done (as a developer) to support a Nautobot major version?
> * What has to be done (as a developer) to support a Nautobot App?
> * What has to be done (as a developer) to support a NetBox new version?
> * What has to be done (as a user) to provide some customization in the import method?

Questions are addressed in `doc/dev/faq.md` and `doc/user/faq.md`.

snaselj commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+- `=>` as a separator.
+- Nautobot field name.
+- Field type in parentheses with the following special cases:
+    - `SKIPPED` means the field is intentionally skipped by the importer.
```
Updated the importer to display disable reason, instead of SKIPPED for each disabled field.

snaselj commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1115 @@
+
+## DiffSync Summary
+
+The first section is the import summary. It shows a summary from the underlying `DiffSync` library.
```
Updated with the link to the DiffSync library.

chadell commented on docs/dev/faq.md:
```diff
@@ -0,0 +1,113 @@
+
+No specific changes to tests should be necessary. The only action required is to cover all deviations between NetBox and Nautobot models in the `nautobot_netbox_importer/diffsync/adapters/<appropriate>.py` file as described above.
+
+## What has to be done to support a major NetBox version?
+
+A new major version of the `nautobot-netbox-importer` package should be released to address updated deviations between NetBox and Nautobot models, independent of the previous major version.
+
+## What has to be done to support a major Nautobot version?
+
+Similar to the previous question, a new independent release of the `nautobot-netbox-importer` package should be released.
```
ok, so every time a new major version for Nautobot or NetBox is released, a new major version of this app has to be released to match it. I guess this will be documented in a versions matrix or similar

snaselj replied:
Seems to be the best option for me. Optionally it would be possible to support multiple major NetBox releases at once. Depends on the release cadence, and the decision which major release matrix should be supported. This can be reconsidered in the future.

Updated compatibility matrix to cover the current status.

snaselj commented on docs/dev/faq.md:
```diff
@@ -0,0 +1,113 @@
+
+Similar to the previous question, a new independent release of the `nautobot-netbox-importer` package should be released.
+
+## What has to be done to support a Nautobot App?
```
Yes, official Nautobot Apps support is a planned feature.

Added Q&A covering Nautobot / NetBox apps support.

snaselj commented:
> * A logical (not technical) diagram for the data from in from NetBox and out to Nautobot
>   

Data flow diagram is here: https://github.com/nautobot/nautobot-app-netbox-importer/pull/110/files#diff-a854f6366a0abe3dea5e2e1ce1802a7123a271bb6670d891c805b923b7cfc77c

bryanculver commented:
I don't pretend to be the one who should be a gatekeeper to this PR, and I'm sure it might take some time to grok the documentation @snaselj has added but I do feel it should be re-reviewed.

snaselj commented on docs/user/summary.md:
```diff
@@ -0,0 +1,1070 @@
+wireless.wirelesslan => wireless.wirelesslan | Disabled with reason: Nautobot content type: `wireless.wirelesslan` not found
+wireless.wirelesslink => wireless.wirelesslink | Disabled with reason: Nautobot content type: `wireless.wirelesslink` not found
+django_rq.queue => django_rq.queue | Disabled with reason: Nautobot content type: `django_rq.queue` not found
+secrets.secret => secrets.secret | Disabled with reason: Nautobot content type: `secrets.secret` not found
```
Added to the follow-up ticket.

chadell reviewed (approved):
LGTM

Let's get this in, and it could be a great opportunity to create a follow-up PR to address the new [NetBox 3.7 release](https://netbox.dev/blog/posts/netbox-v370-released/) to demonstrate how a new version should be added
